[CORL-2384] Upgrade to Node 18

nick-funk opened the pull request:
## What does this PR do?

Upgrades Coral to use Node 18 with npm 8.0.0.

## These changes will impact:

- [ ] commenters
- [ ] moderators
- [ ] admins
- [X] developers

## What changes to the GraphQL/Database Schema does this PR introduce?

None

## Does this PR introduce any new environment variables or feature flags?

No

## If any indexes were added, were they added to `INDEXES.md`?

No new indices.

## How do I test this PR?

- install Node `v18.16.0` with npm `v8.0.0`
  - recommend using `nvm` for this
- run `sh scripts/clear-node_modules.sh`
  - this removes all existing node 14 module files that can cause conflicts
- run `sh npm-i.sh`
  - alternately, run `npm initialize.sh` if you've never run a build before or want to start fresh
- open a terminal in `client/` and run `npm run watch`
- open a terminal in `server/ and run `npm run watch`
- wait for the builds to finish
- navigate to http://localhost:8080 and test out the stream/admin
  - alternatively, do a server only build with `multi-site-test` if that's more to your taste
  
Extra testing:
- create an image using the `Dockerfile`
- spin up the image and see that it runs properly

## Where any tests migrated to React Testing Library?

No

## How do we deploy this PR?

No special considerations.

nick-funk committed:
set engine to node 18 and set compatibility flags

nick-funk committed:
make npm run watch work with node 18

nick-funk committed:
use 127.0.0.1 for amp `npm run watch` pages

nick-funk committed:
Merge branch 'develop' into feat/upgrade-to-node-18

nick-funk committed:
set node 18 options in dockerfile

nick-funk committed:
set node options for server tests

nick-funk committed:
await correct error in tenant protected email domain tests

nick-funk committed:
set max old space size in test to prevent runner overwrite

nick-funk committed:
Revert "set max old space size in test to prevent runner overwrite"

This reverts commit ee677c506361ae347097ce8b9dc99dbfef42740d.

nick-funk committed:
set --no-compilation-cache on jest server tests

suggestion from here: https://github.com/jestjs/jest/issues/11956

nick-funk committed:
set max old space size for gha and jest accordingly

nick-funk committed:
disable cache on server tests

suggested online this could cause the memory leak

nick-funk committed:
upgrade to jest 29 (and everything else that came with it)

nick-funk committed:
set jest workerIdleMemoryLimit=0.2

nick-funk committed:
set workerIdleMemoryLimit in GHA config

nick-funk committed:
remove --no-cache and --runInBand for server tests

suggested here:
https://github.com/jestjs/jest/issues/11956#issuecomment-1225787798

nick-funk committed:
update jest to latest

nick-funk committed:
set worker idle memory limit down to 1024MB

reason:
https://github.com/jestjs/jest/issues/11956#issuecomment-1233105817

nick-funk committed:
set --workerIdleMemoryLimit='1024MB' in testServer.js

nick-funk committed:
try setting back max-old-space-size for terser

nick-funk committed:
check to see if 2 terser threads avoids OOM errors

nick-funk committed:
try 4 terser threads and no source maps for prod builds

nick-funk committed:
disable parallel terser on production builds

nick-funk committed:
outright set parallel to false as terser devs are recommending

nick-funk committed:
remove terser as webpack seems to handle minifying just fine

nick-funk committed:
switch gha runtime to coral-cli over ubuntu-22.04

testing some things with bchan

nick-funk committed:
fix typo coral-cli => coral-ci

nick-funk committed:
increase max-old-space-size to 14 GB

we have 16 GB in the runner now, lets use it

nick-funk committed:
set .nvmrc files to node 18

kabeaty commented on client/config/webpackDevServer.config.ts:
```diff
@@ -69,7 +69,7 @@ export default function ({
     // WebpackDevServer is noisy by default so we emit custom message instead
     // by listening to the compiler events with `compiler.plugin` calls above.
-    quiet: true,
+    quiet: false,
```
Just curious to know why we are setting `quiet` to `false` here now?

nick-funk replied:
just me accidentally leaving this on from debugging watch on node 18, I've set it back to `true`

kabeaty commented on server/package.json:
```diff
@@ -368,10 +368,10 @@
     "terser-webpack-plugin": "^4.2.3",
```
Can we remove this now? And also the `terser-webpack-plugin.d.ts` types file too?

nick-funk replied:
removed!

kabeaty replied:
yay thanks! looks like the types file was removed but not `terser-webpack-plugin` here too?

nick-funk replied:
oh dang, I discarded at some point to try a cleaner approach to something or the other, must have lost removing this when I did that. I've fixed it up and pushed that change up now. Thanks for catching this!

kabeaty reviewed (approved):
Looking great and works as expected! Just had two questions/comments about the changes.

nick-funk committed:
don't try and lint deleted files

nick-funk committed:
remove un-necessary terser plugin types

nick-funk committed:
set quiet back to true now that we're done debugging webpack

nick-funk committed:
Merge branch 'develop' into feat/upgrade-to-node-18

nick-funk committed:
remove terser-webpack-plugin from client package.json

thought I did this earlier, but appaerently I must have
discarded at some point.

tessalt committed:
Merge branch 'develop' into feat/upgrade-to-node-18
